### PR TITLE
[Merged by Bors] - chore: use foldr in List.prod

### PR DESCRIPTION
**Pull request by kim-em**
Makes `List.prod_cons` a `rfl` lemma, without any axioms.

Closes https://github.com/leanprover-community/mathlib3/issues/161

Simplifies some hypotheses, and makes this compatible with `Nat.sum` in core (which I will subsequently remove).

---

- [x] depends on: #17220
- [x] depends on: #17222

### CODE REVIEW

File: Mathlib/Algebra/BigOperators/Group/List.lean [hunk 1]
```diff
@@ -51,21 +51,40 @@ def alternatingProd {G : Type*} [One G] [Mul G] [Inv G] : List G → G
...
+section MulOneClass
+
+variable [MulOneClass M] {l : List M} {a : M}
+
 @[to_additive]
 theorem prod_singleton : [a].prod = a :=
-  one_mul a
+  mul_one a
 
-@[to_additive (attr := simp)]
```
Comment (kim-em) [inline Mathlib/Algebra/BigOperators/Group/List.lean:66]: This no longer needs `@[simp]` as `simp` can already prove it.

File: Mathlib/Algebra/BigOperators/Group/List.lean [hunk 2]
```diff
@@ -51,21 +51,41 @@ def alternatingProd {G : Type*} [One G] [Mul G] [Inv G] : List G → G
...
+@[to_additive]
```
Comment (kim-em) [inline Mathlib/Algebra/BigOperators/Group/List.lean:86]: This no longer needs `simp` and `simp` can prove it.

**Review by YaelDillies — APPROVED** [review]
Looks fine, but could you record the motivation somewhere?

maintainer merge

File: Mathlib/Algebra/BigOperators/Group/List.lean [hunk 3]
```diff
@@ -51,21 +51,41 @@ def alternatingProd {G : Type*} [One G] [Mul G] [Inv G] : List G → G
...
+
+@[to_additive]
+lemma prod_induction
+    (p : M → Prop) (hom : ∀ a b, p a → p b → p (a * b)) (unit : p 1) (base : ∀ x ∈ l, p x) :
+    p l.prod := by
+  induction l with
+  | nil => simpa
+  | cons a l ih =>
+    rw [List.prod_cons]
+    simp only [Bool.not_eq_true, List.mem_cons, forall_eq_or_imp] at base
```
Comment (urkud) [inline Mathlib/Algebra/BigOperators/Group/List.lean:73]: Why do you need something about `Bool` here?

**Comment by urkud** [comment]
Otherwise LGTM. Thanks!
bors d+

**Comment by kim-em** [comment]
bors merge

**Pull request closed without merging**